src/basic/fd-util.c

```diff
@@ -960,6 +960,35 @@ int fd_is_opath(int fd) {
         return FLAGS_SET(r, O_PATH);
 }
 
+int fd_vet_accmode(int fd, int mode) {
+        int flags;
+
+        /* Check if fd is opened with desired access mode.
+         *
+         * Returns > 0 on strict match, == 0 if opened for both reading and writing (partial match),
+         * -EPROTOTYPE otherwise. O_PATH fds are always refused with -EBADFD. */
+
+        assert(fd >= 0);
+        assert((mode & ~O_ACCMODE_STRICT) == 0);
```

Comment:
maybe go further and do `IN_SET(mode, O_RDONLY, O_WRONLY, O_RDWR)`, after all if you mask O_ACC_MODE_STRICT there's a fourth value possible, which is not allowed